stan/math/prim/fun/generalized_inverse.hpp

```diff
@@ -0,0 +1,123 @@
+ in
+ *
+ * <ul><li> Courrieu, Pierre. 2008.  Fast Computation of Moore-Penrose Inverse
+ Matrices.
+ * <i>arXiv</i> <b>0804.4809</b> </li></ul>
+ *
+ * @tparam EigMat type of the matrix (must be derived from \c Eigen::MatrixBase)
+ *
+ * @param G specified matrix
+ * @return Generalized inverse of the matrix (an empty matrix if the specified
+ * matrix has size zero).
+ * @note Because the method inverts a SPD matrix internally that interal matrix
+ may result in small numerical issues that result in a non-SPD error. There are
+ two
+ * <code>generalized_inverse</code> functions, one that uses one input matrix
+ (this one)
+ * and another that works with an input matrix and a small jitter to the
+ diagonal of the internal SPD
+ * matrix.
+ */
+template <typename EigMat, require_eigen_t<EigMat>* = nullptr>
+inline Eigen::Matrix<value_type_t<EigMat>, EigMat::RowsAtCompileTime,
+                     EigMat::ColsAtCompileTime>
+generalized_inverse(const EigMat& G) {
+  using value_t = value_type_t<EigMat>;
+  if (G.size() == 0) {
+    return {};
+  }
+
+  const auto n = G.rows();
+  const auto m = G.cols();
+
+  if (G.rows() == G.cols()) {
+    return inverse(G);
+  }
+
+  if (n < m) {
+    return transpose(mdivide_left_spd(tcrossprod(G), G));
+  } else {
+    return transpose(mdivide_right_spd(G, crossprod(G)));
+  }
+}
+
+/**
+ * Returns the Moore-Penrose generalized inverse of the specified matrix.
+ *
+ * The method is based on the Cholesky computation of the transform as specified
+ in
+ *
+ * <ul><li> Courrieu, Pierre. 2008.  Fast Computation of Moore-Penrose Inverse
+ Matrices.
+ * <i>arXiv</i> <b>0804.4809</b> </li></ul>
+ *
+ * @tparam EigMat type of the matrix (must be derived from \c Eigen::MatrixBase)
+ *
+ * @param G specified matrix
+ * @param a real constant
+ * @return Generalized inverse of the matrix (an empty matrix if the specified
+ * matrix has size zero).
+ * @note Because the method inverts a SPD matrix internally that interal matrix
+ may result in small numerical issues that result in a non-SPD error. There are
+ two
+ * <code>generalized_inverse</code> functions, one that uses one input matrix
+ * and another (this one) that works with an input matrix and a small jitter to
+ the diagonal of the internal SPD
+ * matrix.
+ */
+template <typename EigMat, typename Scal, require_eigen_t<EigMat>* = nullptr,
+          require_stan_scalar_t<Scal>* = nullptr>
+inline Eigen::Matrix<value_type_t<EigMat>, EigMat::RowsAtCompileTime,
+                     EigMat::ColsAtCompileTime>
+generalized_inverse(const EigMat& G, const Scal& a) {
+  using value_t = value_type_t<EigMat>;
+  if (G.size() == 0) {
+    return {};
+  }
+
+  const auto n = G.rows();
+  const auto m = G.cols();
+
+  if (G.rows() == G.cols()) {
+    return inverse(G);
+  }
+
+  if (n < m) {
+    Eigen::Matrix<value_t, Eigen::Dynamic, Eigen::Dynamic> A = tcrossprod(G);
+    A.diagonal().array()
+        += Eigen::Array<double, Eigen::Dynamic, 1>::Constant(n, a);
```

Comment:
[optional] This can be simplified:
```suggestion
    A.diagonal().array() += a;
```
Same for the addition to diagonal in the other branch.

Reply:
Thanks, looks better.